PythonVisualizations/BinaryTreeBase.py

```diff
@@ -0,0 +1,515 @@
+            node.coords[0], node.coords[1] - self.CIRCLE_SIZE)
+      # were the coords passed?
+      elif isinstance(node, tuple):
+            toPos = (node[0], node[1] - self.CIRCLE_SIZE - self.ARROW_HEIGHT,
+            node[0], node[1] - self.CIRCLE_SIZE)
+
+      self.moveItemsTo(arrow, (toPos,), steps = numSteps)
+
+   # calculate the coordinates for the node shape
+   def calculateCoordinates(self, parent, level, childDirection):
+        if level == 0:
+            return self.ROOT_X0, self.ROOT_Y0
+        elif childDirection == Child.LEFT:
+            return parent.coords[0] - 1/2**level* (self.NODE_X_GAP-self.CIRCLE_SIZE), self.ROOT_Y0+level* self.NODE_Y_GAP
+        else:
+            return parent.coords[0] + 1/2**level* (self.NODE_X_GAP-self.CIRCLE_SIZE), self.ROOT_Y0+level* self.NODE_Y_GAP
+
+   # calculate the coordinates for the line attached to the node
+   def calculateLineCoordinates(self, node):
+      x1 = node.coords[0]
+      y1 = node.coords[1] +self.CIRCLE_SIZE
+
+      # where would the child be drawn?
+      level = self.getLevel(node)+1
+      x2, y2 = self.calculateCoordinates(node, level, Child.LEFT)
+      x3, y3 = self.calculateCoordinates(node, level, Child.RIGHT)
+      y2 -= self.CIRCLE_SIZE
+      y3 -= self.CIRCLE_SIZE
+
+      return [(x1, y1, x2, y2), (x1, y1, x3, y3)]
+
+   # highlight the left line of the node
+   def highlightLeftLine(self, node, color = "red"):
+      # get all the items with the node's tag
+      nodeFamily = set(self.canvas.find_withtag(node.tag))
+      # get all the lines that are left
+      lines = set(self.canvas.find_withtag("left line"))
+
+      line = (lines & nodeFamily).pop()
+
+      self.canvas.itemconfig(line, fill=color)
+
+   # unhighlight the left line of the node
+   def unHighlightLeftLine(self, node):
+      # get all the items with the node's tag
+      nodeFamily = set(self.canvas.find_withtag(node.tag))
+      # get all the lines that are left
+      lines = set(self.canvas.find_withtag("left line"))
+
+      line = (lines & nodeFamily).pop()
+
+      self.canvas.itemconfig(line, fill="black")
+
+   # highlight the right line of the node
+   def highlightRightLine(self, node, color="red"):
+      # get all the items with the node's tag
+      nodeFamily = set(self.canvas.find_withtag(node.tag))
+      # get all the lines that are left
+      lines = set(self.canvas.find_withtag("right line"))
+
+      line = (lines & nodeFamily).pop()
+
+      self.canvas.itemconfig(line, fill= color)
+
+   # unhighlight the right line of the node
+   def unHighlightRightLine(self, node):
+      # get all the items with the node's tag
+      nodeFamily = set(self.canvas.find_withtag(node.tag))
+      # get all the lines that are left
+      lines = set(self.canvas.find_withtag("right line"))
+
+      line = (lines & nodeFamily).pop()
+
+      self.canvas.itemconfig(line, fill="black")
```

Comment:
These four methods could be simplified into a single method like `highlightLine(self, node, on=True, child='left')`.  That would make the code shorter and more maintainable.  From the parameters, the appropriate attributes would be applied to the appropriate line (or lines if `child` could be `'both'`).  In another part of the code, I suggest only making the parent link for a new node.  In that case, we could skip the `child` parameter and have a `highlightParentLink()` method.

I also think that the changing the color of the line is not enough to distinguish it in the visualizations.  In addition to color, we should make the width twice or three times wider, or maybe adjust the dash settings.

Reply:
Changes implemented in commit 9937d21